### PR TITLE
Decode main_data in decode_wal_record()

### PR DESCRIPTION
**Pull request by lubennikovaav**

### CODE REVIEW

File: pageserver/src/page_cache.rs [hunk 1]
```diff
@@ -235,6 +235,10 @@ pub struct WALRecord {
     pub lsn: u64, // LSN at the *end* of the record
     pub will_init: bool,
     pub rec: Bytes,
+    // Remember the offset of main_data in rec,
+    // so that we don't have to parse the record again.
+    // If record has no main_data, this offset equals rec.len().
+    pub main_data_offset: usize,
```
Comment (hlinnaka) [inline pageserver/src/page_cache.rs:241]: This seems to be unused, except a debugging `info!()` line.

**Comment by hlinnaka** [comment]
So, if I understand correctly, this doesn't have any user-visible effect yet, but adds infrastructure that will be needed later. Looks good as far as it goes. 

I presume the next step then is to add parsing of commit/abort records drop relations, and SMGR_CREATE and SMGR_TRUNCATE records, so that the relation sizes can be tracked accurately in the page cache. Is that what you have in mind, too?

File: pageserver/src/waldecoder.rs [hunk 2]
```diff
@@ -263,6 +267,7 @@ pub struct DecodedBkpBlock {
     pub rnode_spcnode: u32,
     pub rnode_dbnode: u32,
     pub rnode_relnode: u32,
+    // Note that we have a few special forknum values for non-rel files.
```
Comment (hlinnaka) [inline pageserver/src/waldecoder.rs:270]: A rust enum for the non-rel files would probably be readable.

(I'm still a bit sceptical of having special handling for SLRU files in the page server, as it requires more Postgres changes, but that's another topic)

**Comment by lubennikovaav** [comment]
> So, if I understand correctly, this doesn't have any user-visible effect yet, but adds infrastructure that will be needed later. Looks good as far as it goes.
> 
> I presume the next step then is to add parsing of commit/abort records drop relations, and SMGR_CREATE and SMGR_TRUNCATE records, so that the relation sizes can be tracked accurately in the page cache. Is that what you have in mind, too?

Right. I pushed some more.
It works on top of `request_nonrel` postgres branch and requests all catalog pages and pg_clog pages from pageserver. 
To fix a race caused by walreceiver connection I added a flag - see 4188b73 .
For now, only pg_clog commit replay Is implemented and it seems to be enough to pass the existing test.

The next steps are abort handling, and other record types (subtrans? multixact? smgr?) that we will need. Now, that record parsing and replay infrastructure are ready, it should be easier to implement.

**Pull request merged**